File: python/material.py
from vector3 import *
from util import *
from ray import *
import random

class Material (object):
	def scatter(self, ray_in, hit_record):
		return (False, zero(), zero())

class Lambert (Material):
	def __init__(self, albedo):
		self.albedo = albedo

	def scatter(self, ray_in, hit_record):
		target = hit_record.pt + hit_record.normal + pt_in_unit_sphere()
		scattered = Ray(hit_record.pt, target - hit_record.pt)
		return (True, scattered, self.albedo)

class Metal(Material):
	def __init__(self, albedo, fuzz):
		self.albedo = albedo
		self.fuzz = fuzz

	def scatter(self, ray_in, hit_record):
		r = reflect(ray_in.direction.unit(), hit_record.normal)
		ray = Ray(hit_record.pt, r + pt_in_unit_sphere() * self.fuzz)
		return (ray.direction.dot(hit_record.normal) > 0, ray, self.albedo)

def schlick(cosine, ref_index):
	r0 = (1-ref_index) / (1+ref_index)
	r0 = r0*r0
	return r0 + (1-r0)*math.pow(1-cosine, 0.5)

class Dieletric(Material):
	def __init__(self, ref_index):
		self.ref_index = ref_index

	def scatter(self, ray_in, hit_record):
		# print("in scatter!")
		# print(ray_in)
		# print(hit_record)
		outward_normal = None
		ni_over_nt = None
		cosine = None
		if ray_in.direction.dot(hit_record.normal) > 0:
			outward_normal = -hit_record.normal
			# print(outward_normal)
			ni_over_nt = self.ref_index
			# print(ni_over_nt)
			cosine = ray_in.direction.dot(hit_record.normal) / ray_in.direction.length()
			g = 1.0 - self.ref_index*self.ref_index*(1.0-cosine*cosine)
			cosine = math.sqrt(g) if g > 0 else -99
			# print(g)
			# cosine = math.sqrt(g)
		else:
			outward_normal = hit_record.normal
			ni_over_nt = 1 / self.ref_index
			cosine = -ray_in.direction.dot(hit_record.normal) / ray_in.direction.length()
		refracted, rdir = refract(ray_in.direction, outward_normal, ni_over_nt)
		if (refracted and cosine == -99):
			raise Exception("FUCK")
		reflect_prob = schlick(cosine, self.ref_index) if refracted else 1.0
		attenuation = Vector3(1,1,1)
		if random.random() > reflect_prob:
			# 		print("refracted: " + rdir)
			return (True, Ray(hit_record.pt, rdir), attenuation)
		else:
			reflected = reflect(ray_in.direction, hit_record.normal)
			# print("reflected: " + str(reflected))
			return (True, Ray(hit_record.pt, reflected), attenuation)

File: rust/raytrace/src/material.rs
use std::sync::Arc;

use vec3::Vec3;
use ray::Ray;
use hitable::HitRecord;
use rt_rand::*;
use texture::TexturePtr;

pub struct ScatterInfo {
  pub attenuation: Vec3,
  pub scattered: Ray,
}

pub trait Material {
  // result: attenuation, scatter
  fn scatter(&self, ray: &Ray, hit: &HitRecord) -> Option<ScatterInfo>;
  fn emit(&self, _u: f64, _v: f64, _pt: &Vec3) -> Vec3 {
    Vec3::zero()
  }
}

pub type MaterialPtr = Arc<Material + Sync + Send>;

pub struct Lambertian {
  texture: TexturePtr
}

impl Lambertian {
  pub fn new(texture: TexturePtr) -> Lambertian {
    Lambertian {
      texture
    }
  }

  pub fn rc(texture: TexturePtr) -> Arc<Lambertian> {
    Arc::new(Lambertian::new(texture))
  }
}

impl Material for Lambertian {
  fn scatter(&self, ray: &Ray, hit: &HitRecord) -> Option<ScatterInfo> {
    let target = hit.p + hit.normal + random_in_unit_sphere();
    let scattered = Ray::new(hit.p, target - hit.p, ray.time);
    Some(ScatterInfo {
      attenuation: self.texture.value(hit.u, hit.v, &hit.p),
      scattered
    })
  }
}

pub struct Metal {
  texture: TexturePtr,
  fuzz: f64,
}

impl Metal {
  pub fn new(texture: TexturePtr, fuzz: f64) -> Metal {
    Metal {
      texture,
      fuzz
    }
  }

  pub fn rc(texture: TexturePtr, fuzz: f64) -> Arc<Metal> {
    Arc::new(Metal::new(texture, fuzz))
  }
}

impl Material for Metal {
  fn scatter(&self, ray: &Ray, hit: &HitRecord) -> Option<ScatterInfo> {
    let reflected = Vec3::reflect(&ray.direction.normalized(), &hit.normal);
    let scattered = Ray::new(hit.p, reflected + self.fuzz * random_in_unit_sphere(), ray.time);
    if Vec3::dot(&scattered.direction, &hit.normal) > 0.0 {
      Some(ScatterInfo {
        attenuation: self.texture.value(hit.u, hit.v, &hit.p),
        scattered
      })
    } else {
      None
    }
  }
}

pub struct Dielectric {
  ref_index: f64,
}

impl Dielectric {
  pub fn new(ref_index: f64) -> Dielectric {
    return Dielectric {
      ref_index
    }
  }

  pub fn rc(ref_index: f64) -> Arc<Dielectric> {
    Arc::new(Dielectric::new(ref_index))
  }
}

fn schlick(cosine: f64, ref_index: f64) -> f64 {
  let r0 = (1.0 - ref_index) / (1.0 + ref_index);
  let r2 = r0*r0;
  r2 + (1.0 - r2) * (1.0 - cosine).powf(5.0)
}

impl Material for Dielectric {
  fn scatter(&self, ray: &Ray, hit: &HitRecord) -> Option<ScatterInfo> {
    let reflected = Vec3::reflect(&ray.direction, &hit.normal);
    let ni_over_nt: f64;
    let outward_normal: Vec3;
    let cosine: f64;
    if Vec3::dot(&ray.direction.normalized(), &hit.normal) > 0.0 {
      outward_normal = hit.normal * -1.0;  // todo: unary
      ni_over_nt = self.ref_index;
      let c = Vec3::dot(&ray.direction, &hit.normal) / ray.direction.length();
      let g = 1.0 - self.ref_index*self.ref_index*(1.0 - c*c);
      cosine = if g > 0.0 {
        g.sqrt()
      } else {
        0.0
      };
    } else {
      outward_normal = hit.normal;
      ni_over_nt = 1.0 / self.ref_index;
      cosine = Vec3::dot(&(ray.direction.normalized() * -1.0), &hit.normal.normalized())
    }
    let refracted0 = Vec3::refract(&ray.direction, &outward_normal, ni_over_nt);
    let mut refracted = Vec3::zero();
    let reflect_prob;
    if let Some(r) = refracted0 {
      reflect_prob = schlick(cosine, self.ref_index);
      refracted = r;
    } else {
      reflect_prob = 1.0;
    }
    if rand_f64() < reflect_prob {
      Some(ScatterInfo {
        attenuation: Vec3::one(),
        scattered: Ray::new(hit.p, reflected, ray.time),
      })
    } else {
      Some(ScatterInfo {
        attenuation: Vec3::one(),
        scattered: Ray::new(hit.p, refracted, ray.time)
      })
    }
  }
}

pub struct DiffuseLight {
  texture: TexturePtr,
}

impl DiffuseLight {
  pub fn new(texture: TexturePtr) -> DiffuseLight {
    DiffuseLight {
      texture
    }
  }

  pub fn rc(tex: TexturePtr) -> Arc<DiffuseLight> {
    Arc::new(DiffuseLight::new(tex))
  }
}

impl Material for DiffuseLight {
  fn scatter(&self, _ray: &Ray, _hit: &HitRecord) -> Option<ScatterInfo> {
    None
  }

  fn emit(&self, u: f64, v: f64, pt: &Vec3) -> Vec3 {
    self.texture.value(u, v, pt)
  }
}
